### ISSUE TITLE
Twitter blocks any URL with ? via. robots.txt - Filter twitter URLs to remove "ref_src"

### ISSUE THREAD

**Issue by jerclarke** [berkmancenter/amber_wordpress#44]
Hello and thank you for this wonderful system. 

Our site is enormous, with lots of content each day, and Amber can't keep up with our links. 
We are trying to figure out how to clean up the Amber Dashboard results as best we can to keep things moving (UPDATE: See #47 for details on how the queue woes can be fixed). 

In the process I noticed that almost all our Twitter links (of which there are many, we are a site that reports on citizen media) are in a "Down" state, but some are indeed up. 

What I've discovered is that many/most of the Twitter URLs in our site have an annoying tracker variable embedded in them, like this one (full data from the dashboard):

* twitter.com 
* https://twitter.com/golnarM/status/947469913593131009?ref_src=twsrc%5Etfw 
* Down 
* Sun, 31 Dec 2017 18:26:36 +0000
* Could not capture snapshot

Additionally, looking directly at the database, I see this extra and excruciatingly useful info in the `message` column of the `amber_check` table:

* Blocked by robots.txt

 Sidenote: This info should be surfaced in the dashboard, which just says "could not capture snapshot", which is FAR less useful than the vital information that the URL is flat out banned by robots.txt. I created a ticket for this separate issue: #46

---

To test this, I resaved the post with the link, but having removed `?ref_src=twsrc%5Etfw`. The result was a new entry in the `amber_check` and `amber_cache` tables for the clean version of the URL, this time with UP/1 as the state. 

So clearly:

* Twitter URLs work, but not with `?ref_src=twsrc%5Etfw` style query arguments
* [Twitter's robots.txt](https://twitter.com/robots.txt) is blocking the `?ref_src=twsrc%5Etfw` 
* Amber is wasting it's time trying to submit these URLs which will never work
* Amber is missing out on the opportunity to cache these correctly as tweets someone thought were worth linking to. 

In terms of the robots.txt, this section seems to be what applies:

> \# Every bot that might possibly read and respect this file.
> User-agent: *
> [..]
> Disallow: /*?

If I'm reading this correctly, it means that any Twitter URL with a `?` will get blocked, and is Twitter's way of denying search engines access to equivalent/duplicate URLs, since Twitter never seems to use URL query variables to determine the content that will show. 

---

### EXPECTED BEHAVIOR

IMHO, the most logical behavior for Amber in this context is to **strip URL variables out of Twitter URLs before handling them** in the first placel. The URL variables are ignored by twitter and guarantee that the Amber process will fail, while removing them before fetching/sending will usually fix the problem and result in success.

Alternately, Amber can also choose to discard these URLs completely and not try to fetch them. This is "sad" because it means the URLs won't get preserved, but better than spinning our wheels (my server and archive.org) for no possible utility. 

I realize that these considerations may snowball into a long list of exceptions within Amber, and this is undesirable, but the popularity of Twitter and it's embed system makes Twitter a very important type of URL to not have clogging up our queues. IMHO it's worth solving this directly in the plugin for the sake of all users who will likely be affected. 

### How to solve ourselves?

If this will not be fixed in the core plugin, could you advise on how to safely create this filter ourselves? 

Amber should have some WordPress filters/actions that allow us to accomplish these kind of transformations ourselves, both for common problems like this Twitter one, and for more unique problems (such as a similar issue that might affect our internal intersite linking). 

If this filter already exists, it would be nice to have a reference to it and example of it's use in the documentation. If that existed I could write a patch for my needs very quickly. 

Thanks for your attention and work on this plugin! Hoping for a fruitful relationship going forward :)

**Comment by jerclarke** [comment]
Alright, so in developing our own solution for this we concluded that the only sensible option would be to add two filters to the plugin itself, which we have submitted as PR #48 

These are simple and harmless unless the filter is used, but allow enormous flexibility to developers implementing Amber. 

They have separate filter names (`amber_lookup_url` and `amber_enqueue_check_link`) so that they can be handled separately if necessary, but should usually have the same filtering function/method attached, since for most cases, you'll want both places to do the same manipulations on the URLs. 

In the Twitter case, we can use this filter to remove ?ref_url and any other URL query variables from the URL before it is checked (so the check has  a chance to pass) as well as when the link is checked on display (so that it will match the database entry for that URL, which otherwise would be difference since the database would have the sans-ref_url version but the display checker would have the has-ref_url version). 

This filter can also be used for all kinds of other things too! It could be used to do a very nuanced "blacklist" task that involved checking something more elaborate than Regex, it could be used to completely replace one domain with another etc.

Please, I beg of you, add this filter to the plugin. 

If you like, you can also incorporate the Twitter filter, so that Twitter will work better for all your users. If not that's fine though, having access to the filter is much more important for us.

**Comment by jerclarke** [comment]
NOTE: looking at our error log, I'm seeing a lot of this, despite running the filter described above that avoids triggering `twitter.com/robots.txt`:

```
[28-Jun-2018 15:21:31 UTC] [...]/wp-content/plugins/amberlink/libraries/AmberChecker.php:AmberChecker::check:Blocked by robots.txt:https://t.co/wnn0C8bIVr 
```

Quick look at `http://t.co/robots.txt` explains it: 

```
User-agent: twitterbot
Disallow:

User-agent: *
Disallow: /
```

So it seems any Twitter link in the `t.co` format will always die, which is a bummer. 

Possibility to fix: Have a filter (like the one above) that does the full permalink expansion of `t.co` URLs before doing anything with them, so that the system only gets the working `twitter.com` version. 

Haven't done it yet but documenting it here for myself and/or others.
